### PR TITLE
[1.6] Add support for Python 3.11, drop support for Python 3.7

### LINKED ISSUE #3561: Uprev pinned dependencies in `develop-1.6` where possible
### Proposed Changes

Review dependencies in `develop` that have newer versions than what we're currently pinned to, and update to any new ones that don't have breaking changes that impact us. Similar to updates made in `next` in #3554.

### Justification

Keeping LTM from getting too stale.

### PR DESCRIPTION
**Pull request by glennmatthews**
In addition to explicitly adding Python 3.11, this is basically a backport of #3554 and #3027 and #3674 to `develop-1.6` branch, with a few additional dependency updates to latest versions made possible by and/or required for 3.11 compatibility.

Closes #3561, though I haven't exhaustively checked all of our dependencies to see what else we can uprev - just the latest versions of the same set of dependencies that were touched in those three PRs.

# What's Changed

- Lots of dependency updates, including some major version bumps (I've reviewed the changelogs a bit and didn't see any red flags, but am mostly relying on testing and CI to catch any breakage)
- Lots of Python file formatting changes due to upreving black (virtually identical to #3554)
- A few more substantive Python file changes due to upreving pylint (virtually identical to #3027)
- Updates to workflows, etc. to drop 3.7 (virtually identical to #3554) and then add 3.11 (new)

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- n/a Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by bryanculver** [comment]
Can we also uprev `napalm`.

**Comment by u1735067** [comment]
If that can be useful, `celery<5.3` (specifically `billiard<4.1.0`) is affected by https://github.com/celery/billiard/issues/377 with Python 3.11, so you might want to upgrade Celery too (one way to trigger this issue is by having a job that fails in a hard way, like the DB going down when writing a log entry).

**Comment by glennmatthews** [comment]
Thank you @u1735067! I'll look into updating celery, it looks like it should be feasible here.

File: nautobot/core/tests/test_releases.py [hunk 1]
```diff
@@ -99,6 +99,7 @@ def test_pre_releases(self, mock_cache_get: Mock, mock_cache_set: Mock, mock_req
             "https://localhost/unittest/releases",
             headers={"Accept": "application/vnd.github.v3+json"},
             proxies=settings.HTTP_PROXIES,
+            timeout=15,
```
Comment (HanlinMiao) [inline nautobot/core/tests/test_releases.py:102]: Just for my own curiosity, why did we add the time out here?

**Review by HanlinMiao — APPROVED** [review]
Looks good going through the list of dependency changes and upgrades!

File: nautobot/core/tests/test_releases.py (hunk 1 above)
Comment (glennmatthews) [reply nautobot/core/tests/test_releases.py:102]: Complaint by pylint that `requests` can potentially hang forever if there's no timeout provided. We already did the same change in `next` in #3027.

**Pull request merged**